docs/design/podaffinity.md

```diff
@@ -0,0 +1,615 @@
+if a pod P requests anti-affinity, P cannot schedule onto a node with conflicting pods,
+and pods that conflict with P cannot schedule onto the node one P has been scheduled there.
+The design we have described says that the symmetry property for RequiredDuringScheduling *affinity*
+is weaker: if a pod P says it can only schedule onto nodes running pod Q, this
+does not mean Q can only run on a node that is running P, but the scheduler will try
+to schedule Q onto a node that is running P (i.e. treats the reverse direction as
+preferred). This raises the same scheduling quality concern as we menioned at the
+end of the Denial of Service section above, and can be addressed in similar ways.
+
+The nature of affinity (as opposed to anti-affinity) means that there is no issue of
+determining which pod(s) to kill
+when a pod's labels change: it is obviously the pod with the affinity rule that becomes
+violated that must be killed. (Killing a pod never "fixes" violation of an affinity rule;
+it can only "fix" violation an anti-affinity rule.) However, affinity does have a
+different question related to killing: how long should the system wait before declaring
+that RequiredDuringSchedulingRequiredDuringExecution affinity is no longer met at runtime?
+For example, if a pod P has such an affinity for a pod Q and pod Q is temporarily killed
+so that it can be updated to a new binary version, should that trigger killing of P? More
+generally, how long should the system wait before declaring that P's affinity is
+violated? (Of course affinity is expressed in terms of label selectors, not for a specific
+pod, but the scenario is easier to describe using a concrete pod.) This is closely related to
+the concept of forgiveness (see issue #1574). In theory we could make this time duration be
+configurable by the user on a per-pod basis, but for the first version of this feature we will
+make it a configurable property of whichever component does the killing and that applies across
+all pods using the feature. Making it configurable by the user would require a nontrivial change
+to the API syntax (since the field would only apply to RequiredDuringSchedulingRequiredDuringExecution
+affinity).
+
+## Implementation plan
+
+1. Add the `Affinity` field to PodSpec and the `PodAffinity` and `PodAntiAffinity` types to the API along with all of their descendant types.
+2. Implement a scheduler predicate that takes `RequiredDuringSchedulingIgnoredDuringExecution`
+affinity and anti-affinity into account. Include a workaround for the issue described at the end of the Affinity section of the Examples section (can't schedule first pod).
+3. Implement a scheduler priority function that takes `PreferredDuringSchedulingIgnoredDuringExecution` affinity and anti-affinity into account
+4. Implement a quota mechanism that charges for the entire topology domain when `RequiredDuringScheduling` anti-affinity is used. Later
```

Comment:
Yes, that's reasonable. I think what we'd specifically want is: "prohibit specifying 'all namespaces' with non-'node' TopologyKey for RequiredDuringScheduling anti-affinity? (I think we don't want to prohibit "exclusive machine".)